### PR TITLE
Add top layer for gui

### PR DESCRIPTION
**Pull request by rinkk**
added the functionality of @TomFischer 's AddTopLayer tool to "normal" OGS/GUI
layers can now also be added at bottom of mesh

### CODE REVIEW

**Comment by bilke** [comment]
Jenkins failure is due to Petsc related failing tests, everything else is fine.

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 1]
```diff
@@ -114,85 +62,10 @@ int main (int argc, char* argv[])
...
-		mesh_arg.getValue().c_str());
-	const MathLib::Vector3 dir(0,0,-1);
-	double const angle(90);
-	std::unique_ptr<MeshLib::Mesh> sfc_mesh(
-		MeshLib::MeshSurfaceExtraction::getMeshSurface(
-			*subsfc_mesh, dir, angle, true
-		)
-	);
-	INFO("done.");
+	MeshLib::Mesh* result (MeshLib::addTopLayerToMesh(*subsfc_mesh, layer_thickness_arg.getValue(), mesh_out_arg.getValue()));
```
Comment (endJunction) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: Please break all long lines.

File: Applications/DataExplorer/DataView/MshView.cpp [hunk 2]
```diff
@@ -180,13 +186,37 @@ void MshView::openValuesEditDialog()
 	valueEdit.exec();
 }
 
+void MshView::openAddLayerDialog()
+{
+	QModelIndex const index = this->selectionModel()->currentIndex();
```
Comment (endJunction) [inline Applications/DataExplorer/DataView/MshView.cpp]: Could you briefly explain why `this->` is needed?

File: Applications/DataExplorer/DataView/MshView.cpp [hunk 3]
```diff
@@ -157,8 +163,8 @@ void MshView::contextMenuEvent( QContextMenuEvent* event )
 
 void MshView::openMeshEditDialog()
 {
-	MshModel* model = static_cast<MshModel*>(this->model());
-	QModelIndex index = this->selectionModel()->currentIndex();
+	MshModel const*const model = static_cast<MshModel*>(this->model());
+	QModelIndex const index = this->selectionModel()->currentIndex();
 	const MeshLib::Mesh* mesh =
 	        static_cast<MshModel*>(this->model())->getMesh(index);
```
Comment (TomFischer) [inline Applications/DataExplorer/DataView/MshView.cpp]: Is it possible to reuse the already fetched `model` (line 166) here?

File: Applications/DataExplorer/DataView/MshView.cpp [hunk 4]
```diff
@@ -170,8 +176,8 @@ void MshView::openMeshEditDialog()
 
 void MshView::openValuesEditDialog()
 {
-	MshModel* model = static_cast<MshModel*>(this->model());
-	QModelIndex index = this->selectionModel()->currentIndex();
+	MshModel const*const model = static_cast<MshModel*>(this->model());
+	QModelIndex const index = this->selectionModel()->currentIndex();
 	MeshLib::Mesh* mesh = const_cast<MeshLib::Mesh*>(static_cast<MshModel*>(this->model())->getMesh(index));
```
Comment (TomFischer) [inline Applications/DataExplorer/DataView/MshView.cpp]: Same as above: Can the `model` be reused here?

File: Applications/Utils/MeshEdit/AddTopLayer.cpp (hunk 1 above)
Comment (TomFischer) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: Maybe it is possible to use `std::unique_ptr` here?

File: MeshLib/MeshEditing/AddLayerToMesh.cpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
+/**
+ * \file   AddLayerToMesh.cpp
+ * \author Kasten Rink
```
Comment (TomFischer) [inline MeshLib/MeshEditing/AddLayerToMesh.cpp]: Do you really want to be the author of my buggy code? :smile:

File: Applications/DataExplorer/DataView/MshView.cpp [hunk 6]
```diff
@@ -202,23 +232,23 @@ void MshView::extractSurfaceMesh()
...
-	QSettings settings;
-	QFileInfo fi (settings.value("lastOpenedMeshFileDirectory").toString());
-	const MeshLib::Mesh* mesh = static_cast<MshModel*>(this->model())->getMesh(index);
-	QString fileName = QFileDialog::getSaveFileName(NULL, "Convert mesh to shapefile...",
-		                                    LastSavedFileDirectory::getDir() + QString::fromStdString(mesh->getName()),
-											"ESRI Shapefile (*.shp)");
+	QSettings const settings;
+	QFileInfo const fi (settings.value("lastOpenedMeshFileDirectory").toString());
+	MeshLib::Mesh const*const mesh = static_cast<MshModel*>(this->model())->getMesh(index);
+	QString const fileName = QFileDialog::getSaveFileName(NULL, "Convert mesh to shapefile...",
```
Comment (TomFischer) [inline Applications/DataExplorer/DataView/MshView.cpp]: `NULL`-> `nullptr`

File: Applications/Utils/MeshEdit/AddTopLayer.cpp (hunk 1 above)
Comment (rinkk) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: VtuInterface wants a "regular" pointer. I added a delete for the mesh instead.

File: Applications/DataExplorer/DataView/MshView.cpp (hunk 2 above)
Comment (rinkk) [inline Applications/DataExplorer/DataView/MshView.cpp]: "this" is not strictly needed but I use it to indicate class variables and methods of the Qt-GUI classes where our normal naming conventions don't apply or it's difficult to look up the definitions.

**Comment by rinkk** [comment]
@TomFischer Not sure if you would like to expand your cmd line utility. The final version now supports adding layers above and below the mesh and I also added support for 1d/2d meshes, i.e. you can expand a polyline into a curtain now and things like that...

**Comment by endJunction** [comment]
Looks good. If you need some help for rebasing and squashing afterwards, do not hesitate to ask me or Tom.
:fast_forward:

**Pull request merged**